### DISCUSSION TITLE
Missing TearDown listeners for laravel echo

### DISCUSSION THREAD

**Discussion by walirazzaq** [livewire/livewire#4335] [3. Feature ideas]
Current implementation of laravel-echo events does not have `tearDown` listeners and in case of component is remove from livewire store, echo-listeners stay hanging around there.
Normally it doesn't break anything but when are component is removed and added back conditionally or if using some navigation libraries such as `Turbo`. 
Echo listeners keeps on adding for components. When the echo event is fired in this situation livewire reacts to that but it contains multiple `fireEvent` actions under `updates`. 
Following is the example payload for a component listening to an echo event. I am using `Turbo` for SPA navigation everytime moving to the page having a such component listening for echo events adds extra entry to `payload updates`.
This can be reproduced without `Turbo` by conditionally rendering livewire component multiple times. Livewire is handling this in single `ajax` call though but `payload ` size keeps on increasing.
```JSON
{
    "fingerprint": {...},
    "serverMemo": {...},
    "updates": [
        {
            "type": "fireEvent",
            "payload": {
                "id": "ytij",
                "event": "echo-private:App.Customer.1,.CustomerUpdated",
                "params": [...]
            }
        },
        {
            "type": "fireEvent",
            "payload": {
                "id": "le7n",
                "event": "echo-private:App.Customer.1,.CustomerUpdated",
                "params": [...]
            }
        },
        {
            "type": "fireEvent",
            "payload": {
                "id": "ikub",
                "event": "echo-private:App.Customer.1,.CustomerUpdated",
                "params": [...]
            }
        }
    ]
}
```
One solution to this can be adding `tearDown` callbacks to `component` when handling `laravel-echo`. For proof of concept if something like this can be done
```JS
/**
 * https://github.com/livewire/livewire/blob/master/js/component/LaravelEcho.js
 */
if (['channel', 'private', 'encryptedPrivate'].includes(channel_type)) {
    Echo[channel_type](channel).listen(event_name, e => {
        store.emit(event, e)
    })

    component.addListenerForTeardown(() => {
        Echo[channel_type](channel).stopListening(event_name)
    })

    /** OR */
    
    component.addListenerForTeardown(() => {
        Echo.leaveChannel(channel)
    })
}
```
Leaving the channel seems obvious as it will `unsubscribe` the channel itself and other components which have subscribed to the same channel before will not be getting updates.

I am not sure about `stopListening` though if it works the same way.

Cases for `presense` channels and `notification` is a bit different. For `.notification(...)`, `laravel-echo` is using hard coded event name
```JS
    /**
    * https://github.com/laravel/echo/blob/30a41d1599d2aec5303d0e0f6bf2aef790f1cf1f/src/channel/channel.ts#L26
     */
    notification(callback: Function): Channel {
        return this.listen('.Illuminate\\Notifications\\Events\\BroadcastNotificationCreated', callback);
    }

```
If `.stopListening(...)` works as I am perceiving it to be, same can be done for `.notification(...)`.
And for `presence` channels I am still digging to understand it's working.

Thank You

**Comment by joshhanley** [comment]
@walirazzaq yeah good call, it should definitely remove the listeners if the component is removed/ torn down.

If you have a solution for that, I'm sure PR's would be welcome! 🙂

**Reply by walirazzaq** [reply]
Hi @joshhanley 
I tried the above solution it does remove the listeners but it does it for all. I mean suppose there are two components listening for same event. One of them got torn down and it's Echo listener is also removed. The other component which is listening to same event would stop working (listening) as well.

For now I am unable to find a proper way which doesn't disturb other components when torn down

**Reply by joshhanley** [reply]
@walirazzaq ah hmm yeah that's not ideal. I haven't looked into how they are setup so can't be of much help just yet. But I'm planning to review all the websockets/ echo support in the new year so I can have a look then 🙂
